Start work on custom autonomous P2P trajectories

MathNerd28 opened the pull request:
- Resolve slight merge issues
- Change camera location to front of test chassis
- Filter target readings with high ambiguity
- Add cosine wheel direction matching to swerve drive
- Expose additional methods in swerve drive

MathNerd28 committed:
Start work on custom autonomous P2P trajectories

- Resolve slight merge issues
- Change camera location to front of test chassis
- Filter target readings with high ambiguity
- Add cosine wheel direction matching to swerve drive
- Expose additional methods in swerve drive

MathNerd28 commented:
I’d like to pretty much scrap yesterday’s work with waypoints, and instead go for a more generalized approach as seen in [this post on Chief Delphi](https://www.chiefdelphi.com/t/paper-implementation-of-the-adaptive-pure-pursuit-controller/166552)

MathNerd28 committed:
Implement swerve trajectory generation

- Create classes and commands to generate autonomous trajectories with independent headings
- Remove unused SDS swerve library (now using custom replacement)
- Slight refactoring

MathNerd28 committed:
Begin designing routes + simulation

mmarchetti commented on src/main/java/frc/lib/SwerveTrajectory.java:
```diff
@@ -0,0 +1,87 @@
+public class SwerveTrajectory {
+    public static final double MAXIMUM_LINEAR_VELOCITY      = 0.7
+            * Constants.Swerve.MAXIMUM_LINEAR_VELOCITY;
+    public static final double MAXIMUM_LINEAR_ACCELERATION  = MAXIMUM_LINEAR_VELOCITY;
```
This implies an acceleration time of 1 second. You might consider making that explicit by making a constant for the accel time and dividing by it here. (Doesn't have to be part of this PR).

mmarchetti commented on src/main/java/frc/lib/SwerveTrajectory.java:
```diff
@@ -0,0 +1,87 @@
+    public static final double MAXIMUM_LINEAR_ACCELERATION  = MAXIMUM_LINEAR_VELOCITY;
+    public static final double MAXIMUM_ANGULAR_VELOCITY     = 0.5
+            * Constants.Swerve.MAXIMUM_ANGULAR_VELOCITY;
+    public static final double MAXIMUM_ANGULAR_ACCELERATION = MAXIMUM_ANGULAR_VELOCITY;
```
Same here

mmarchetti commented on src/main/java/frc/lib/SwerveTrajectoryGenerator.java:
```diff
@@ -0,0 +1,279 @@
+
+            double velocityChange = predictedVelocities[i]
+                    - predictedVelocities[i - 1];
+            double timeChange = times[i] - times[i - 1];
```
If you were to name these variables `dv` and `dt` it will look a lot like calculus.

mmarchetti commented on src/main/java/frc/lib/SwerveTrajectoryGenerator.java:
```diff
@@ -0,0 +1,279 @@
+        for (int i = 0; i < deltas.length; i++) {
+            deltas[i][0] = waypoints[i + 1].getX() - waypoints[i].getX();
+            deltas[i][1] = waypoints[i + 1].getY() - waypoints[i].getY();
+            deltas[i][2] = waypoints[i + 1].getRotation()
```
In this function you might consider whether assigning some temporaries variables to hold expressions like `waypoints[i+1]` or later `points[pointIndex]` might help readability.

mmarchetti commented on src/main/java/frc/robot/AutoRoutes.java:
```diff
@@ -0,0 +1,128 @@
+    public static final Pose2d[] START_3_TO_PIECE_4 = { START_POINT_3, S3_P4_A,
+            S3_P4_B, S3_P4_C, S3_P4_D, S3_P4_E, GAME_PIECE_4 };
+    public static final Pose2d[] PIECE_4_TO_START_3 = { GAME_PIECE_4, S3_P4_E,
+            S3_P4_D, S3_P4_C, S3_P4_B, S3_P4_A, START_POINT_3 };
```
It's good to have a start at an auto route to see the trajectory generation working. Let's let Doug and Cameron take it from here.

mmarchetti commented on src/main/java/frc/robot/commands/SwerveTrajectoryCommand.java:
```diff
@@ -0,0 +1,132 @@
+        return delta.getTranslation()
+                    .getNorm() < 0.05
+                && Math.abs(delta.getRotation()
+                                 .getDegrees()) < 2;
```
The position and angle tolerances should probably be named constants.

mmarchetti reviewed (approved):
I left a few comments which you can address in this PR or merge as-is.

MathNerd28 committed:
Make requested PR changes + cleanup

- Move all constants to new Constants member class
- Implement SwerveTrajectory::toTrajectory explicitly without Stream API
- Rename variables to read more clearly
- Remove/replace unused SwerveModule kP, kD, kF constants
